src/build_options.rs

```diff
@@ -725,6 +691,166 @@ impl BuildOptions {
+    for dep in main_dependencies {
+        let dist_name = get_distribution_name(dep.clone())
+            .context("Your dependencies table in pyproject.toml is invalid")?;
+        main_deps_names.insert(dist_name, dep.to_string());
+    }
+    debug!("main dependencies {:?}", main_deps_names);
+
+    // distribution name -> (requires string, crate name)
+    let mut metadata_deps_names: HashMap<String, Vec<(String, String)>> = HashMap::new();
+    // I've manually tested that this doesn't include optional dependencies
+    for package in &metadata.packages {
+        let dependencies = package
+            .metadata
+            .get("pyo3-packaging")
+            // also called dependencies after https://peps.python.org/pep-0621/#dependencies-optional-dependencies
+            .and_then(|pyo3_metadata| pyo3_metadata.get("dependencies"))
+            .and_then(|dependencies| {
+                serde_json::from_value::<Vec<String>>(dependencies.clone()).ok()
+            })
+            .unwrap_or_default();
+        if !dependencies.is_empty() {
+            debug!("crate {} got dependencies {:?}", package.name, dependencies);
+        }
+
+        for dep in dependencies {
+            let dist_name = get_distribution_name(dep.clone())
+                .context("Your dependencies table in pyproject.toml is invalid")?;
+            if let Some(main_requires) = main_deps_names.get(&dist_name) {
+                // Ideally, with full PEP440 support, we could merge all those ourselves and just
+                // notify the user like "added {} and merged {} python dependencies from the
+                // metadata of cargo dependencies"
+                println!(
+                    "➡️ Replacing '{}' from crate {} with pyproject.toml '{}'",
+                    dep, package.name, main_requires
+                );
+                continue;
+            }
+            let requires_and_source = metadata_deps_names
+                .entry(dist_name.to_string())
+                .or_default();
+            // If the two crates use the exact same requires string
+            if !requires_and_source
+                .iter()
+                .any(|(requires, _cargo_package)| requires == &dep)
+            {
+                requires_and_source.push((dep.clone(), package.name.clone()));
+            }
+            merged_requires.push(dep.clone());
+        }
+    }
+    debug!("metadata dependencies {:?}", metadata_deps_names);
+
+    for (dep_name, sources) in metadata_deps_names {
+        if sources.len() > 1 {
+            let joined = sources
+                .iter()
+                .map(|(requires, cargo_package)| {
+                    format!("'{}' in crate {}", requires, cargo_package)
+                })
+                .collect::<Vec<String>>()
+                .join(", ");
+            println!(
+                "⚠️ Warning: {dep_name} is specified multiple times: {joined}. \
```

Comment:
To be honest, I think this is serious downside. If the issue was very common, i.e. users being swamped with managing the transitive Python dependencies of the Rust dependencies, it might make sense to handle at least the common case.

But since I assume that this is rare (Do we have any example besides NumPy which is somewhat special by exposing a capsule-based C-API for us to consume?), I am not to convince that carrying this without full PEP440 support and limited dependency resolution is really worth it.

Reply:
> But writing a proper PEP440 parser is on my list anyway, so maybe i'll do that

Maybe this ties in with https://github.com/PyO3/pyproject-toml-rs or https://github.com/PyO3/python-pkginfo-rs? Or could be organized as a similar project/crate?

Reply:
yep absolutely!